### PR TITLE
Introducing MNE Deep

### PR DESCRIPTION
**Pull request by chdinh**

### CODE REVIEW

File: MNE/deep/deep.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+{
+
+//*************************************************************************************************************
+//=============================================================================================================
+// FORWARD DECLARATIONS
+//=============================================================================================================
+
+//=============================================================================================================
+/**
+* BEM descritpion
```
Comment (LorenzE) [inline MNE/deep/deep.h]: please check docu

File: MNE/deep/deep.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+*
+* @brief BEM descritpion
```
Comment (LorenzE) [inline MNE/deep/deep.h]: see above

File: mne-cpp.pri [hunk 3]
```diff
@@ -34,13 +34,59 @@ QMAKE_TARGET_DESCRIPTION = MNE Qt 5 based C++ library.
...
+## To build basic MNE-X version run: qmake MNECPP_CONFIG+=BuildBasicMNESCANVersion
+#MNECPP_CONFIG += BuildBasicMNESCANVersion
+
+## Build MNE-CPP libraries as static libs
+#MNECPP_CONFIG += build_MNECPP_Static_Lib
+
+## Build MNE-CPP Deep library
+MNECPP_CONFIG += buildDeep
+
+#Build minimalVersion for qt versions <5.7.0
```
Comment (LorenzE) [inline mne-cpp.pri]: pls change this to 5.7.1 due to qt3d

File: MNE/deep/deep.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+#include "deep.h"
+
+#ifdef _WIN32
+#include <Windows.h>
+#endif
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// STL INCLUDES
```
Comment (LorenzE) [inline MNE/deep/deep.cpp]: We are not using STL include section anymore. Only INLCUDES, Qt INCLUDES, Eigen INCLUDES

File: MNE/deep/deep.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+#include "deep_global.h"
+
+#include <Eval.h>
+//#include <CNTKLibrary.h>
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// FIFF INCLUDES
```
Comment (LorenzE) [inline MNE/deep/deep.h]: same as above

File: MNE/deep/deep.h [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+//=============================================================================================================
+
+
+
+//*************************************************************************************************************
+//=============================================================================================================
+// STL INCLUDES
```
Comment (LorenzE) [inline MNE/deep/deep.h]: same as above

File: MNE/deep/deep.h [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    * Returns the Output Dimensions
+    *
+    * @return the Output dimensions
+    */
+    size_t outputDimensions();
+
+private:
+    QString m_sModelFilename;                               /**< Model filename */
+
+    Microsoft::MSR::CNTK::IEvaluateModel<float>* m_model;   /**< The loaded model */
```
Comment (LorenzE) [inline MNE/deep/deep.h]: please rename to m_pModel and use forward declarations

File: MNE/deep/deepeval.cpp [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+void DeepEval::setModelFilename(const QString &sModelFilename)
+{
+    m_sModelFilename = sModelFilename;
+}
+
+
+//*************************************************************************************************************
+
+bool DeepEval::evalModel(std::vector<float>& inputs, std::vector<float>& outputs)
```
Comment (LorenzE) [inline MNE/deep/deepeval.cpp]: use const for input?

File: MNE/deep/deepeval.h [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+* @brief    DeepEval class declaration.
+*
+*/
+
+#ifndef DEEPEVAL_H
+#define DEEPEVAL_H
+
+//*************************************************************************************************************
+//=============================================================================================================
+// MNE INCLUDES
```
Comment (LorenzE) [inline MNE/deep/deepeval.h]: Just call this INCLUDES and put everything in here except for qt and eigen

File: MNE/deep/deepeval.h [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    * Returns the Output Dimensions
+    *
+    * @return the Output dimensions
+    */
+    size_t outputDimensions();
+
+private:
+    QString m_sModelFilename;                               /**< Model filename */
+
+    Microsoft::MSR::CNTK::IEvaluateModel<float>* m_model;   /**< The loaded model */
```
Comment (LorenzE) [inline MNE/deep/deepeval.h]: rename to m_pModel and use forward declaration.

File: examples/ex_deep/ex_deep.pro [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+LIBS += -L$${CNTK_LIBRARY_DIR}
+CONFIG(debug, debug|release) {
+    LIBS += -lMNE$${MNE_LIB_VERSION}Genericsd \
+            -lMNE$${MNE_LIB_VERSION}Utilsd \
+            -lMNE$${MNE_LIB_VERSION}Fsd \
+            -lMNE$${MNE_LIB_VERSION}Fiffd \
+            -lMNE$${MNE_LIB_VERSION}Mned \
+            -lMNE$${MNE_LIB_VERSION}Inversed \
+            -lMNE$${MNE_LIB_VERSION}Deepd \
+            -lEvalDll
```
Comment (LorenzE) [inline examples/ex_deep/ex_deep.pro]: does a debug version for lEvalDll exist?

File: examples/ex_deep_eval/ex_deep_eval.pro [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+LIBS += -L$${CNTK_LIBRARY_DIR}
+CONFIG(debug, debug|release) {
+    LIBS += -lMNE$${MNE_LIB_VERSION}Genericsd \
+            -lMNE$${MNE_LIB_VERSION}Utilsd \
+            -lMNE$${MNE_LIB_VERSION}Fsd \
+            -lMNE$${MNE_LIB_VERSION}Fiffd \
+            -lMNE$${MNE_LIB_VERSION}Mned \
+            -lMNE$${MNE_LIB_VERSION}Inversed \
+            -lMNE$${MNE_LIB_VERSION}Deepd \
+            -lEvalDll
```
Comment (LorenzE) [inline examples/ex_deep_eval/ex_deep_eval.pro]: debug?

File: examples/ex_deep_eval/main.cpp [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+{
+    QApplication a(argc, argv);
+
+    DeepEval deepTest(QApplication::applicationDirPath() + "/mne_deep_models/examples/output/models/ex_deep_one_hidden");
+
+    //
+    // Generate dummy input values in the appropriate structure and size
+    //
+    std::vector<float> inputs, outputs;
+    for (int i = 0; i < deepTest.inputDimensions(); i++)
```
Comment (LorenzE) [inline examples/ex_deep_eval/main.cpp]: please use brackets, even for one liners

File: examples/ex_deep_eval/main.cpp [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        inputs.push_back(static_cast<float>(i % 255));
+
+    deepTest.evalModel(inputs, outputs);
+
+    //
+    // Output the results
+    //
+    for (auto& value : outputs)
```
Comment (LorenzE) [inline examples/ex_deep_eval/main.cpp]: please use brackets, even for one liners

File: examples/ex_deep_eval/main.cpp [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+        fprintf(stderr, "%f\n", value);
+
+    //
+    // Visualize
+    //
+    QLineSeries *series = new QLineSeries();
+    for (int i = 0; i < outputs.size(); i++)
```
Comment (LorenzE) [inline examples/ex_deep_eval/main.cpp]: please use brackets, even for one liners

File: MNE/deep/deep.h (hunk 7 above)
Comment (chdinh) [inline MNE/deep/deep.h]: Forward declaration with template classes do not exist.

File: MNE/deep/deepeval.cpp (hunk 8 above)
Comment (chdinh) [inline MNE/deep/deepeval.cpp]: i'm afraid a CNTK method used inside doe only take non const - thats the reason while it is here also not const

File: MNE/deep/deepeval.h (hunk 10 above)
Comment (chdinh) [inline MNE/deep/deepeval.h]: done

File: examples/ex_deep/ex_deep.pro (hunk 11 above)
Comment (chdinh) [inline examples/ex_deep/ex_deep.pro]: i am afraid, it does not exists

File: examples/ex_deep_eval/ex_deep_eval.pro (hunk 12 above)
Comment (chdinh) [inline examples/ex_deep_eval/ex_deep_eval.pro]: nope

File: examples/ex_deep_eval/main.cpp (hunk 13 above)
Comment (chdinh) [inline examples/ex_deep_eval/main.cpp]: done

File: examples/ex_deep_eval/main.cpp (hunk 14 above)
Comment (chdinh) [inline examples/ex_deep_eval/main.cpp]: done

File: examples/ex_deep_eval/main.cpp (hunk 15 above)
Comment (chdinh) [inline examples/ex_deep_eval/main.cpp]: done

**Pull request merged**